storage-proofs/porep/src/nse/circuit/circuit.rs

```diff
@@ -21,6 +38,471 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> Circuit<Bls12>
     for NseCircuit<'a, Tree, G>
 {
     fn synthesize<CS: ConstraintSystem<Bls12>>(self, cs: &mut CS) -> Result<(), SynthesisError> {
-        todo!()
+        let Self {
+            replica_id,
+            comm_r,
+            comm_d,
+            layer_proofs,
+            comm_layers,
+            public_params,
+            ..
+        } = self;
+
+        // Allocate replica_id
+        let replica_id_num = num::AllocatedNum::alloc(cs.namespace(|| "replica_id"), || {
+            replica_id
+                .map(Into::into)
+                .ok_or_else(|| SynthesisError::AssignmentMissing)
+        })?;
+
+        // make replica_id a public input
+        replica_id_num.inputize(cs.namespace(|| "replica_id_input"))?;
+
+        // get the replica_id in bits
+        let replica_id_bits =
+            reverse_bit_numbering(replica_id_num.to_bits_le(cs.namespace(|| "replica_id_bits"))?);
+
+        // comm_d
+        // Allocate comm_d as Fr
+        let comm_d_num = num::AllocatedNum::alloc(cs.namespace(|| "comm_d"), || {
+            comm_d
+                .map(Into::into)
+                .ok_or_else(|| SynthesisError::AssignmentMissing)
+        })?;
+
+        // make comm_d a public input
+        comm_d_num.inputize(cs.namespace(|| "comm_d_input"))?;
+
+        // Allocate comm_r as Fr
+        let comm_r_num = num::AllocatedNum::alloc(cs.namespace(|| "comm_r"), || {
+            comm_r
+                .map(Into::into)
+                .ok_or_else(|| SynthesisError::AssignmentMissing)
+        })?;
+
+        // make comm_r a public input
+        comm_r_num.inputize(cs.namespace(|| "comm_r_input"))?;
+
+        // Allocate comm_layers
+        let comm_layers_nums = comm_layers
+            .into_iter()
+            .enumerate()
+            .map(|(i, comm)| {
+                num::AllocatedNum::alloc(cs.namespace(|| format!("comm_layer_{}", i)), || {
+                    comm.map(Into::into)
+                        .ok_or_else(|| SynthesisError::AssignmentMissing)
+                })
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut comm_layers_nums_padded = comm_layers_nums.clone();
+        let arity = PoseidonMDArity::to_usize();
+        while comm_layers_nums_padded.len() % arity != 0 {
+            comm_layers_nums_padded.push(num::AllocatedNum::alloc(
+                cs.namespace(|| format!("padding_{}", comm_layers_nums_padded.len())),
+                || Ok(Fr::zero()),
+            )?);
+        }
+
+        // Verify hash(comm_layers) == comm_r
+        {
+            let hash_num = PoseidonFunction::hash_md_circuit::<_>(
+                &mut cs.namespace(|| "comm_layers_hash"),
+                &comm_layers_nums_padded,
+            )?;
```

Comment:
BLOCKING: Whoops, per Poseidon audit, I think we shouldn't be using the Poseidon HashMD function as-is. We can probably just use the specific arity needed (is it 14, 15?) instead. Alternately, if we need variable length (which I don't think is the case here), we can use one of the other constructions the auditors proposed — one of which might include a small tweak to the extant MD function. I don't have access to the document at the moment, but will look into this and deal with it next week.

Reply:
> Whoops, per Poseidon audit, I think we shouldn't be using the Poseidon HashMD function as-is.

:(  that is unfortunate, yes the hashing is a specific arity, but we do vary that in the tests quite a bit currently. But technically we can fix it to either a hardcoded number, or simply add a method which uses an existing artiy up to the one we need.

Reply:
I'll look at the options and try to implement one of those which accommodates variable length. I think the easiest version is not bad, and the other might be worth doing too. Understood about the tests — makes sense.

Reply:
@dignifiedquire And I discussed this, with the resolution that we will change to use a fixed arity hash equal to the value actually required for real circuits (15). When fewer layers exist because an insecure circuit is used for faster/easier testing, we will hardcode the padding in the circuit.

Reply:
After more discussion with Poseidon auditors, I will probably end up implementing a different scheme for this. That will require calling a new function here. @dignifiedquire I think you should proceed as planned, and when the new API function is available, I will replace the usage in rust-fil-proofs.